datasets/trecqa.py

```diff
@@ -26,29 +31,63 @@ def __init__(self, path):
 
     @classmethod
     def splits(cls, path, train='train-all', validation='raw-dev', test='raw-test', **kwargs):
         return super(TRECQA, cls).splits(path, train=train, validation=validation, test=test, **kwargs)
 
     @classmethod
-    def iters(cls, path, vectors_name, vectors_cache, batch_size=64, shuffle=True, device=0, vectors=None, unk_init=torch.Tensor.zero_):
+    def set_vectors(cls, field, vector_path):
+        if os.path.isfile(vector_path):
+            stoi, vectors, dim = torch.load(vector_path)
+            field.vocab.vectors = torch.Tensor(len(field.vocab), dim)
+
+            for i, token in enumerate(field.vocab.itos):
+                wv_index = stoi.get(token, None)
+                if wv_index is not None:
+                    field.vocab.vectors[i] = vectors[wv_index]
+                else:
+                    # initialize <unk> with U(-0.25, 0.25) vectors
+                    field.vocab.vectors[i] = torch.FloatTensor(dim).uniform_(-0.05, 0.05)
+        else:
+            print("Error: Need word embedding pt file")
+            print("Error: Need word embedding pt file")
+            exit(1)
+        return field
+
+    @classmethod
+    def iters(cls, path, vectors_name, vectors_dir, batch_size=64, shuffle=True, device=0, pt_file = False, vectors=None, unk_init=torch.Tensor.zero_):
         """
         :param path: directory containing train, test, dev files
         :param vectors_name: name of word vectors file
-        :param vectors_cache: directory containing word vectors file
+        :param vectors_dir: directory containing word vectors file
         :param batch_size: batch size
         :param device: GPU device
         :param vectors: custom vectors - either predefined torchtext vectors or your own custom Vector classes
         :param unk_init: function used to generate vector for OOV words
         :return:
         """
-        if vectors is None:
-            vectors = Vectors(name=vectors_name, cache=vectors_cache, unk_init=unk_init)
 
         train, validation, test = cls.splits(path)
+        if not pt_file:
+            if vectors is None:
+                vectors = Vectors(name=vectors_name, cache=vectors_dir, unk_init=unk_init)
+            cls.TEXT_FIELD.build_vocab(train, validation, test, vectors=vectors)
+        else:
+            cls.TEXT_FIELD.build_vocab(train, validation, test)
+            cls.TEXT_FIELD = cls.set_vectors(cls.TEXT_FIELD, os.path.join(vectors_dir, vectors_name))
+
+        cls.LABEL_FIELD.build_vocab(train, validation, test)
 
-        cls.TEXT_FIELD.build_vocab(train, validation, test, vectors=vectors)
+        cls.VOCAB_NUM = len(cls.TEXT_FIELD.vocab)
 
-        return BucketIterator.splits((train, validation, test), batch_size=batch_size, repeat=False, shuffle=shuffle, device=device)
+        # BucketIterator
+        return Iterator.splits((train, validation, test), batch_size=batch_size, repeat=False, shuffle=shuffle, device=device)
```

Comment:
Why are we changing from `BucketIterator` to `Iterator`?

`BucketIterator` "minimizes amount of padding needed while producing freshly shuffled batches for each new epoch."

Reply:
You are right. I use Iterator because NCE will select samples so padding is necessary. Then I suppose during the test stage there should be also padding. I will run an experiment to check it tonight. I will add a bool variable to choose the Iterator/BucketIterator if it is true. Otherwise, I will change to BucketIterator only.